client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -0,0 +1,264 @@
+
+    @Override
+    public void add(Uri uri, Cookie cookie) {
+        String thisRequestDomain = requestDomain(uri);
+        String thisRequestPath = requestPath(uri);
+
+        try {
+            lock.lock();
+            add(thisRequestDomain, thisRequestPath, cookie);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public List<Cookie> get(Uri uri) {
+        try {
+            lock.lock();
+            return get(requestDomain(uri), requestPath(uri), uri.isSecured());
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public List<Cookie> getAll() {
+        try {
+            lock.lock();
+            final boolean[] removeExpired = {false};
+            List<Cookie> result = cookieJar
+                    .entrySet()
+                    .parallelStream()
+                    .filter(pair -> {
+                        boolean hasCookieExpired = hasCookieExpired(pair.getValue().cookie, pair.getValue().createdAt);
+                        if (hasCookieExpired && !removeExpired[0])
+                            removeExpired[0] = true;
+                        return !hasCookieExpired;
+                    })
+                    .map(pair -> pair.getValue().cookie)
+                    .collect(Collectors.toList());
+
+            if (removeExpired[0])
+                removeExpired();
+
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean remove(Cookie cookie) {
+        try {
+            lock.lock();
+            return cookieJar.entrySet().removeIf(v -> v.getValue().cookie == cookie);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean clear() {
+        try {
+            lock.lock();
+            boolean result = !cookieJar.isEmpty();
+            cookieJar.clear();
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    private String requestDomain(Uri requestUri) {
+        return requestUri.getHost().toLowerCase();
+    }
+
+    private String requestPath(Uri requestUri) {
+        if (!requestUri.getPath().isEmpty())
+            return requestUri.getPath();
+        else
+            return "/";
+    }
+
+    // rfc6265#section-5.2.3
+    // Let cookie-domain be the attribute-value without the leading %x2E (".") character.
+    private AbstractMap.SimpleEntry<String, Boolean> cookieDomain(String cookieDomain, String requestDomain) {
+        if (cookieDomain != null) {
+            if (!cookieDomain.isEmpty() && cookieDomain.charAt(0) == '.')
+                return new AbstractMap.SimpleEntry<>(cookieDomain.substring(1).toLowerCase(), false);
+            else
+                return new AbstractMap.SimpleEntry<>(cookieDomain.toLowerCase(), false);
```

Comment:
Compute normalized cookie domain once with a ternary operator and use one single return statement

Reply:
Normalized how ? It's converted to lower case only. Is this something you're looking for ?

```
private AbstractMap.SimpleEntry<String, Boolean> cookieDomain(String cookieDomain, String requestDomain) {
    if (cookieDomain != null) {
        String normalizedCookieDomain = cookieDomain.toLowerCase();
        return (!cookieDomain.isEmpty() && cookieDomain.charAt(0) == '.') ?
                new AbstractMap.SimpleEntry<>(normalizedCookieDomain.substring(1), false) :
                new AbstractMap.SimpleEntry<>(normalizedCookieDomain, false);
    } else
        return new AbstractMap.SimpleEntry<>(requestDomain, true);
}
```